internal/app/machined/pkg/controllers/k8s/apiserver_endpoints.go

```diff
@@ -0,0 +1,185 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/channel"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/config/machine"
+	"github.com/siderolabs/talos/pkg/machinery/resources/cluster"
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/k8s"
+)
+
+// APIServerEndpointsController creates a list of API server endpoints.
+type APIServerEndpointsController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Name() string {
+	return "cluster.APIServerEndpointsController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.To(config.MachineTypeID),
+			Kind:      controller.InputWeak,
+		},
+		safe.Input[*cluster.Member](controller.InputWeak),
+		safe.Input[*config.MachineConfig](controller.InputWeak),
+		safe.Input[*cluster.Identity](controller.InputWeak),
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.APIServerEndpointsType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *APIServerEndpointsController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		if _, ok := channel.RecvWithContext(ctx, r.EventCh()); !ok && ctx.Err() != nil {
+			return nil //nolint:nilerr
+		}
+
+		machineConfig, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine config: %w", err)
+			}
+
+			continue
+		}
+
+		machineType, err := safe.ReaderGetByID[*config.MachineType](ctx, r, config.MachineTypeID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine type: %w", err)
+			}
+
+			continue
+		}
+
+		identity, err := safe.ReaderGetByID[*cluster.Identity](ctx, r, cluster.LocalIdentity)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting local identity: %w", err)
+			}
+
+			continue
+		}
+
+		members, err := safe.ReaderListAll[*cluster.Member](ctx, r)
```

Comment:
should we conditionally check this based on if discovery is enabled (kubespan/k8s registry)?

Reply:
I'm not sure. If discovery is disabled, we will get an empty list and will simply skip adding from that list later on.